In [1]:
# Import packages
import pandas as pd
import numpy as np
import os
os.getcwd()

'/home/jovyan/PS'

# Lists (6pt)

    1. (2pt) Create a list ‘movies‘ that contains the names of at least six movies you like

In [2]:
# Make a list of six movie names
movies = ["Lord of the Rings", "Hacksaw Ridge", "Avengers", "Hobbits", "King's Speech", "Saving Private Ryan"]

    2. (2pt) Use slicing, and create a list of three first movies in the list

In [3]:
# Select indices from the list
first_three_movies = movies[0:3]
print(first_three_movies)

['Lord of the Rings', 'Hacksaw Ridge', 'Avengers']


    3. (2pt) Use slicing and list concatenation to create a list of the first two and the last two movies

In [4]:
# Use slicing and list concatenation to get a list of first two and last two elements of the movie list
first_last_two_movies = movies[:2] + movies[-2:]
print(first_last_two_movies)

['Lord of the Rings', 'Hacksaw Ridge', "King's Speech", 'Saving Private Ryan']


# Loops (6pt)

    1. (2pt) Create a list ‘numbers‘ that is the numbers 70 through 79 in the following manner: create an empty list and add numbers to it in a loop (do not use list comprehension or ‘list‘ function here!)

In [5]:
# Create an empty list
numbers = []
# Create for loop than runs from 70 and stops before 80 and put values in the list
for i in range(70, 80):
        numbers.append(i)
print(numbers)

[70, 71, 72, 73, 74, 75, 76, 77, 78, 79]


    2. (2pt) Use loop to compute sum of all integers from 1 till 100.

In [6]:
# Set initial sum to 0
sum_int_100 = 0
# Create for loop that runs from 1 and stops before 100 and add the values
for i in range(1,101):
    sum_int_100 += i
print(sum_int_100)

5050


    3. (2pt) Assign people to seats. Consider names Adam, Ashin, Inukai, Tanaka, and Ikki; and seats 33, 12, 45, 2 and 17. Print seat assignments by assigning each name to the corresponding seat. You coude should output
    Adam: 33
    Ashin: 12
    ...
    Hint: loop over the integer range of the length of names (here 0..4, remember–python indexing is
    0-based), and use indexing to access the corresponding name and seat number.

In [7]:
# Create an empty list
seat_assignment = []
# Assign names and seat numbers to two lists
seats = [33, 12, 45, 2, 17]
names = ["Adam", "Ashin", "Inukai", "Tanaka", "Ikki"]
# For loop from 0 and stops before 5 and combine names and seats together
for i in range(0, len(names)):
    print(names[i] + ": " + str(seats[i]))

Adam: 33
Ashin: 12
Inukai: 45
Tanaka: 2
Ikki: 17


# Comprehensions (4pt)

    1. (2pt) Use list comprehension to create a list of squares of numbers 1..10 (i.e. 1, 4, 9, ..., 100)

In [8]:
# Create an empty list
square_list = []
# Append values into the list by using a loop
for i in range(1,11):
  square_list.append(i ** 2)
print(square_list)

[1, 4, 9, 16, 25, 36, 49, 64, 81, 100]


    2. (2pt) Use dict comprehension to create a dict of numbers 1..10 and their squares (i.e. 1:1, 2:4, ..., 10:100).

In [9]:
# Create an empty dictionary
square_dict = {}
# For loop to calculate and insert values into the dictionary
for i in range(1, 11):
    square_dict[i] = i ** 2
print(square_dict)

{1: 1, 2: 4, 3: 9, 4: 16, 5: 25, 6: 36, 7: 49, 8: 64, 9: 81, 10: 100}


# Functions and data transformations (8pt)

    1. (8pt) Write a function that takes in time in the numeric form (i.e. not a string) of HHMM (hoursminutes), and returns it in the numeric form of HH.HH (hours + fractions of hours). For instance, 1015 → 10.25 (10 hrs 15 mins → 10.25 hrs). Demonstrate it works using values 1015 and 345. The function should return (not print) the result as a number. Hint: use modulo operator % and integer division operator //. Modulo of 100 gives you minutes and integer division by 100 gives you hours

In [10]:
# Function: time_converter, convert a time in numeric form and returns in the numeric form of HH.HH
# Input: initial_time, must be numeric
# Output: returns the final converted time 
def time_converter(initial_time):
    minute = initial_time % 100
    hour = initial_time // 100
    fraction = minute / 60
    final_time = hour + fraction
    return(final_time)
# Test using 345 and 1015
test1_345 = time_converter(345)
print(test1_345)
test2_1015 = time_converter(1015)
print(test2_1015)


3.75
10.25


# Load data (8pt)

    1. (2pt) Load the csv data file

In [11]:
# Load gapminder data
gapminder_data = pd.read_csv('/home/jovyan/PS/data/gapminder.csv.bz2', sep='\t')

    2. (6pt) Do basic sanity checks:

    (a) How many variables (columns) is there in the data? Ensure you know the variables in the
    data. Keep the documentation nearby.

In [12]:
# Find columns using shape()
num_columns = gapminder_data.shape[1]
print('Number of columns =', num_columns)

Number of columns = 25


    (b) How many rows of data is there?

In [13]:
# Find rows using shape()
num_rows = gapminder_data.shape[0]
print('Number of rows =', num_rows)

Number of rows = 13055


    (c) print the first few lines of data. Does it look reasonable?

In [14]:
# Print the first 10 lines of the data
gapminder_data.head(10)

iso3   name iso2    region                       sub-region  \
0  ABW  Aruba   AW  Americas  Latin America and the Caribbean   
1  ABW  Aruba   AW  Americas  Latin America and the Caribbean   
2  ABW  Aruba   AW  Americas  Latin America and the Caribbean   
3  ABW  Aruba   AW  Americas  Latin America and the Caribbean   
4  ABW  Aruba   AW  Americas  Latin America and the Caribbean   
5  ABW  Aruba   AW  Americas  Latin America and the Caribbean   
6  ABW  Aruba   AW  Americas  Latin America and the Caribbean   
7  ABW  Aruba   AW  Americas  Latin America and the Caribbean   
8  ABW  Aruba   AW  Americas  Latin America and the Caribbean   
9  ABW  Aruba   AW  Americas  Latin America and the Caribbean   

  intermediate-region    time  totalPopulation  fertilityRate  lifeExpectancy  \
0           Caribbean  1960.0          54211.0          4.820          65.662   
1           Caribbean  1961.0          55438.0          4.655          66.074   
2           Caribbean  1962.0          56225.0          4.471          66.444   
3           Caribbean  1963.0          56695.0          4.271          66.787   
4           Caribbean  1964.0          57032.0          4.059          67.113   
5           Caribbean  1965.0          57360.0          3.842          67.435   
6           Caribbean  1966.0          57715.0          3.625          67.762   
7           Caribbean  1967.0          58055.0          3.417          68.095   
8           Caribbean  1968.0          58386.0          3.226          68.436   
9           Caribbean  1969.0          58726.0          3.054          68.784   

   ...  accessElectricity  agriculturalLand  agricultureTractors  \
0  ...                NaN               NaN                  NaN   
1  ...                NaN              20.0                  NaN   
2  ...                NaN              20.0                  NaN   
3  ...                NaN              20.0                  NaN   
4  ...                NaN              20.0                  NaN   
5  ...                NaN              20.0                  NaN   
6  ...                NaN              20.0                  NaN   
7  ...                NaN              20.0                  NaN   
8  ...                NaN              20.0                  NaN   
9  ...                NaN              20.0                  NaN   

   cerealProduction  fertilizerHa        co2  greenhouseGases     co2_PC  \
0               NaN           NaN  11092.675              NaN  204.62037   
1               NaN           NaN  11576.719              NaN  208.82281   
2               NaN           NaN  12713.489              NaN  226.11808   
3               NaN           NaN  12178.107              NaN  214.80037   
4               NaN           NaN  11840.743              NaN  207.61578   
5               NaN           NaN  10623.299              NaN  185.20396   
6               NaN           NaN   9933.903              NaN  172.11995   
7               NaN           NaN  12236.779              NaN  210.77907   
8               NaN           NaN  11378.701              NaN  194.88749   
9               NaN           NaN  14891.687              NaN  253.57911   

   pm2.5_35  battleDeaths  
0       NaN           NaN  
1       NaN           NaN  
2       NaN           NaN  
3       NaN           NaN  
4       NaN           NaN  
5       NaN           NaN  
6       NaN           NaN  
7       NaN           NaN  
8       NaN           NaN  
9       NaN           NaN  

[10 rows x 25 columns]

The data looks reasonable to me. Although the first 10 lines of the dataframe has some NA values for the country Aruba, there are values for other countries under the categories such as accessElectricity.

# Wealth (36pt)

    First, let’s do some data exploration. Answer the following questions: show the code, the computation
    results, and comment the results in the accompanying text. Remember: you have to explain your results
    as appropriate! Your task is not just to get correct answers but to convince us that you understand what
    you are doing.

    1. (2pt) How many different countries are there in the data?

In [15]:
# Use unique to find the number of different countries in the data
num_countries = gapminder_data['name'].nunique()
print('There are', num_countries, 'different countries in the data.')

There are 249 different countries in the data.


    2. (1pt) What is the earliest and the most recent year in the dataset?

In [16]:
# Use min/max to comput the earliest and the most recent year in the data
print(gapminder_data['time'].min())
print(gapminder_data['time'].max())

1960.0
2019.0


The earlist year in the dataset is 1960. The most recent year in the dataset is 2019.

    Now let’s define wealth as GDP per capita and let’s explore countries by average wealth.

    3. (2pt) For which year do we have the most recent GDP data?
    Hint: You can remove all cases where GDP is missing.

In [17]:
# Drop the NA values from the GDP_PC column and use max to find the year of the most recent data
time_gdp = gapminder_data[["GDP_PC","time"]].dropna()
time_gdp.time.max()

2019.0

We have the most recent GDP data in 2019. 

    4. (3pt) What is the average wealth on this planet as of 2019? Let’s just compute average GDP across
    all countries in 2019 and ignore the fact that countries are of different size.

In [18]:
# Set condition and calculate GDP mean
avg_wealth = time_gdp[time_gdp.time == 2019.0]['GDP_PC'].mean()
print('The average wealth on this planet as of 2019 is:', avg_wealth)

The average wealth on this planet as of 2019 is: 15055.352217243239


    5. (6pt) But not all countries may have the same most recent year where GDP data is present. Which 5
    countries countries have the largest number of the most recent years missing? (For instance, imagine
    there is GDP data for Funan for 2015, 2016, and 2017; but for Khmer Empire only for 2015. Hence
    Khmer Empire has two of the most recent years missing.) Till which year do these countries have
    data? What do you think, why do these countries have issues with more recent data?
    Hint: you may group by country and find max value for the year. In the resulting series, find the
    min/max. Check out the nlargest method.
    Hint2: two of these countries are Lichtenstein and Faroe Islands.

In [19]:
# Create another dataframe 
gdp_data = gapminder_data[['name', 'time', 'GDP_PC']]
# Drop the NA values 
gdp_data = gdp_data[~gdp_data.GDP_PC.isna()]
# Group by country names and find the max value for the year
gdp_data = gdp_data.groupby(['name']).max()
# Find the 5 smallest time
gdp_data.nsmallest(5, 'time')

time        GDP_PC
name                                          
Djibouti                  2010.0    1343.26873
Faroe Islands             2010.0   48530.23385
Liechtenstein             2010.0  141200.38001
Turks and Caicos Islands  2010.0   21028.40784
Eritrea                   2011.0     681.24637

For the 5 countries that have the largest number of the most recent years missing, Djibouti, Faroe Islands, Liechtenstein, Turks, and Caicos Islands have data untill 2010. Eritrea has data until 2011. These 5 countries have issue with more recent data might be caused by factors such as their locations, size of the country. For islands like Faroe Islands, and Turks and Caico Islands, their locations make the collectors very hard to collect their data. Also, the population there can be very low and thus makes the GDP_PC less important for analysis. For countries like Djibouti, LIechtenstein, and Eritrea, their political system seem to be dictatorship. Most of the time, dictatorship also means that the government system is broken and no one was really collecting the data. These might be some reasons why these countries do not have recent years data.

    6. (6pt) Now let’s compare the continents. We’ll make it easy again and just compute the average
    wealth (i.e. GDP) for each continent in 2019, and we use region as continent. We disregard the fact
    that countries are of different size. Print the continents, and the corresponding GDP in a decreasing
    order. Do you think this order is reasonable?
    Remember to use only the most recent data!
    Hint: check out methods groupby and sort_values.
    Hint2: you should see value around 11,800 for Oceania

In [20]:
# Filter the year to 2019, group the data by region, calculate the mean,then sort the values in descending value.
group_region = gapminder_data[(gapminder_data.time == 2019.0)].groupby(gapminder_data.region).mean().sort_values('GDP_PC', ascending=False)
# Show only the region, time, and GDP information.
group_region[['time', 'GDP_PC', 'totalPopulation']]

time        GDP_PC  totalPopulation
region                                         
Europe    2019.0  33227.983767     1.654336e+07
Americas  2019.0  14516.826303     2.249392e+07
Asia      2019.0  14116.124163     9.084813e+07
Oceania   2019.0  11817.735554     2.234829e+06
Africa    2019.0   2765.708956     2.458150e+07

I think the order is reasonable. There are many developed countries in Europe which makes them wealthier. The Americas has countries like Canada and the United States which brings up the GDP a lot. However, there are still many developing countries which is the reason why the GDP of Americas is not as wealthy as Europe. For the last three, there are not many developed countries and many countries are still developing which makes sense. Population can also be a factor. The larger a country becomes the more the economically non-contributing population also increases. Smaller countries can control their economy better. Europe has relatively small population compares to other countries besides Oceania. It is interesting to see how Asia has such a massive population but is still able to be in the third place in terms of GDP.

    7. (6pt) But this was just about the average numbers. Now for each continent let’s also find the richest
    and the poorest country, the corresponding GDP, and population (2019).
    Note: While this gives a hint about inequality, we still ignore the intra-county inequality. Quite
    likely the rich in the poor countries earn more than the poor in the rich countries. Unfortunately
    we cannot tell based on these data.
    Hint: if you can find a good solution yourself, go for it! But here is a suggestion how you can do it,
    step-by-step.
    (a) Check out indexing with .loc and .iloc (McKinney, p 143)
    (b) Check out methods .idxmax and .idxmin (McKinney p 158-159)
    (c) Find the index of the largest GDP value in data (use .idxmax).
    (d) Extract the row of dataframe that corresponds to the index you computed in 7c
    (e) Now repeat the two previous steps either with .groupby, or if you cannot figure this out then
    with a for-loop over all continents.
    Alternatively, you can extract the values using construct like data.gdp == data.gdp.min(). You
    can also just loop over continents, and for each continent find the richest and poorest country as of
    2019 (check out methods nlargest and nsmallest).

In [21]:
# Set the year to 2019, group by region, find the largest GDP value in each continent group
gapminder_data[(gapminder_data.time == 2019.0)].groupby('region').apply(lambda x: x.nlargest(1, 'GDP_PC'))[['name', 'GDP_PC', 'totalPopulation']]

name        GDP_PC  totalPopulation
region                                                   
Africa   11249  Seychelles   15048.74693          97625.0
Americas 1505      Bermuda   94106.36125          64027.0
Asia     9924        Qatar   63281.67468        2832067.0
Europe   7039   Luxembourg  111043.53170         620001.0
Oceania  663     Australia   57183.37935       25365745.0

In [22]:
# Set the year to 2019, group by region, find the smallest GDP value in each continent group
gapminder_data[(gapminder_data.time == 2019.0)].groupby('region').apply(lambda x: x.nsmallest(1, 'GDP_PC'))[['name', 'GDP_PC', 'totalPopulation']]

name      GDP_PC  totalPopulation
region                                                      
Africa   843            Burundi   208.07473       11530580.0
Americas 5176             Haiti  1245.01250       11263077.0
Asia     119        Afghanistan   573.28755       38041754.0
Europe   12211          Ukraine  3224.93517       44386203.0
Oceania  10408  Solomon Islands  1749.42726         669823.0

The richest countries are: Seychelles(Africa), Bermuda(Americas), Qatar(Asia), Luxembourg(Europe), Australia(Oceania).

The poorest countries are: Burundi(Africa), Haiti(Americas), Afghanistan(Asia), Ukraine(Europe), Solomon Islands(Oceania).

    8. (6pt) Finally, let’s put all this into a single data frame. The data frame should contain continent
    name as index, and for each continent 6 variables: name, GDP and population for the poorest
    country, and the same for the richest country. Pick suitable variable names!
    The result might look something like:
    poorest GDP population richest GDP population
    region
    Africa Burundi 208.07473 11530580.0 Seychelles 15048.74693 97625.0
    ...
    Hint: consult Python notes 4.6: Combining data into data frames. You can use .set_index to
    convert a variable into index.

In [23]:
# Create poor and rich variables for easy use in the later data frame
poor = gapminder_data[(gapminder_data.time == 2019.0)].groupby('region').apply(lambda x: x.nsmallest(1, 'GDP_PC')).set_index("region")
rich = gapminder_data[(gapminder_data.time == 2019.0)].groupby('region').apply(lambda x: x.nlargest(1, 'GDP_PC')).set_index("region")
# Create two dataframes, one with the GDP and population of poor countries. One with GDP and population of rich countries.
df_poor = pd.DataFrame({'Poorest': poor['name'], 'GDP': poor['GDP_PC'], 'Population': poor['totalPopulation']})
df_rich = pd.DataFrame({'Richest': rich['name'], 'GDP': rich['GDP_PC'], 'Population': rich['totalPopulation']})
# Combine the two dataframes together
pd.concat([df_poor, df_rich], axis = 1)

Poorest         GDP  Population     Richest           GDP  \
region                                                                        
Africa            Burundi   208.07473  11530580.0  Seychelles   15048.74693   
Americas            Haiti  1245.01250  11263077.0     Bermuda   94106.36125   
Asia          Afghanistan   573.28755  38041754.0       Qatar   63281.67468   
Europe            Ukraine  3224.93517  44386203.0  Luxembourg  111043.53170   
Oceania   Solomon Islands  1749.42726    669823.0   Australia   57183.37935   

          Population  
region                
Africa       97625.0  
Americas     64027.0  
Asia       2832067.0  
Europe      620001.0  
Oceania   25365745.0

    9. (4pt) Comment the list of poorest and richest countries. What do you think about these lists. Did
    you know that Bermuda is the richest country in Americas? Do you know why? Why do most of
    the rich countries have small population?

It is very interesting to these two lists. It is very surprising to see most of these countries becasue I know most of the poor countries but I do not recognize any of the rich countries besides Australia. It seems like most of the poor countries have a massive amount of population while the rich countries have less population. I definitely did not know Bermuda is the richest country in Americas. According to Wikipedia, Bermuda is primarily fueled by offshore financial services for non-resident firms, such as offshore insurance and tourism. It is also likely that the GDP of Bermuda is raised by some of the rich people there. Most people can still live in poverty. Countries with smaller population tends to have higher GDP because they have built up a sel-sufficient economy based on an abundance of special resources. Generally, these countries have small population but makes a lot of money by selling their resources which increase their GDP. Also, the larger a country becomes the more the economically non-contributing population also increases. Smaller countries can control their economy better.

# Health (32pt)

    This is a more demanding problem. Health is a complex concept, but fortunately we can proxy health
    with life expectancy (LE). It is a natural index of health that has been measured rather well for decades,
    or even centuries

    1. (3pt) How many countries do not have LE data for year 1960? How many countries do not have
    this information for year 2019?

In [24]:
# Use unique to find the countries that do not have life expectancy data in year 1960 and 2019
num_uniq_1960 = gapminder_data[(gapminder_data.time == 1960.0) & (gapminder_data.lifeExpectancy.isna())]['name'].nunique()
num_uniq_2019 = gapminder_data[(gapminder_data.time == 2019.0) & (gapminder_data.lifeExpectancy.isna())]['name'].nunique()
print('The are', num_uniq_1960, 'countries (including countries that have NA values) that do not have life expectancy data in the year of 1960. The are', num_uniq_2019, 'countries (including countries that have NA values) that do not have life expectancy data in the year of 2019.)')

The are 24 countries (including countries that have NA values) that do not have life expectancy data in the year of 1960. The are 17 countries (including countries that have NA values) that do not have life expectancy data in the year of 2019.)


    2. (4pt) What is the shortest and longest LE in data? Which years/countries does this correspond to?
    Hint: you can use methods nlargest to extract dataframe rows, and idxmax to extract index of
    row with the maximum value.

In [25]:
# Find the longest life expectancy and its corresponding country and name
gapminder_data.nlargest(1, 'lifeExpectancy')[["name", "time", "lifeExpectancy"]]

name    time  lifeExpectancy
10581  San Marino  2012.0        85.41707

In [26]:
# Find the shortest life expectancy and its corresponding country and name
gapminder_data.nsmallest(1, 'lifeExpectancy')[["name", "time", "lifeExpectancy"]]

name    time  lifeExpectancy
6097  Cambodia  1977.0          18.907

The shortest life expectancy in the data is about 18.907 in Cambodia, 1977. The longest life expectancy in the data is about 85.42 in San Marino, 2012.

    3. (4pt) If you did this correctly, you notice that the shortest LE is less than 20 years. What historical
    events does it correspond to? (You may consult Wikipedia).

In 1977, Cambodia was ruled by the Khmer Rouge and Khmer Rouge carried out the Cambodian genocide from 1975 until 1979. The Cambodian genocide was the systematic persecution and killing of Cambodians by the Khmer Rouge. During that time period, Cambodia lost about a third of its population due to starvation, forced labour, and execution. That is the reason why life expectancy was only 18.9 in 1977.

    4. (6pt) Find the country with longest and shortest LE for each continent.
    Hint: you can use the same approach as for GDP by continent.

In [27]:
# Group by region and find the longest life expectancy in each continent.
gapminder_data[(gapminder_data.time == 2019.0)].groupby('region').apply(lambda x: x.nlargest(1, 'lifeExpectancy'))[['name', 'lifeExpectancy']]

name  lifeExpectancy
region                                    
Africa   3310      Algeria        76.88000
Americas 1986       Canada        82.04878
Asia     4995    Hong Kong        85.07805
Europe   2107  Switzerland        83.70488
Oceania  663     Australia        82.90000

In [28]:
# Group by region and find the shortest life expectancy in each continent.
gapminder_data[(gapminder_data.time == 2019.0)].groupby('region').apply(lambda x: x.nsmallest(1, 'lifeExpectancy'))[['name', 'lifeExpectancy']]

name  lifeExpectancy
region                                                  
Africa   1926   Central African Republic        53.28300
Americas 5176                      Haiti        64.00100
Asia     119                 Afghanistan        64.83300
Europe   12211                   Ukraine        71.82732
Oceania  9445           Papua New Guinea        64.50100

In Africa, the country with the longest life expectancy is Algeria, the country with the shortest life expectancy is Cnetral African Republic.
In Americas, the country with the longest life expectancy is Canada, the country with the shortest life expectancy is Haiti.
In Asia, the country with the longest life expectancy is Hong Kong, the country with the shortest life expectancy is Afghanistan.
In Europe, the country with the longest life expectancy is Switzerland, the country with the shortest life expectancy is Ukraine.
In Oceania, the country with the longest life expectancy is Australia, the country with the shortest life expectancy is Papua New Guinea.

    5. (4pt) Now lets find, for each country, the first valid LE. Find the corresponding year, and LE value.
    Hint: you can use the same approach as when finding the largest GDP on each continent: find the
    index of minimum year of each country, and then extract the rows based on the index.


In [29]:
# Find the indices where lifeExpectncy is first valid for each country
idx_min = gapminder_data[gapminder_data.lifeExpectancy.notna()].groupby('name').time.idxmin()
# Create a sub data frame where life expectancy data is first available
valid_df = gapminder_data.loc[idx_min].set_index("name")
# Show the corresponding country, year, and life expectancy
valid_df[['time', 'lifeExpectancy']]

time  lifeExpectancy
name                                         
Afghanistan            1960.0        32.44600
Albania                1960.0        62.28300
Algeria                1960.0        46.14100
Angola                 1960.0        37.52400
Antigua and Barbuda    1960.0        61.96800
...                       ...             ...
Viet Nam               1960.0        59.03900
Virgin Islands (U.S.)  1960.0        66.22485
Yemen                  1960.0        29.91900
Zambia                 1960.0        46.68700
Zimbabwe               1960.0        53.01900

[204 rows x 2 columns]

    6. Now repeat this with the most recent valid LE.

In [30]:
# Find the indices where lifeExpectncy is most recent valid for each country
idx_max = gapminder_data[gapminder_data.lifeExpectancy.notna()].groupby('name').time.idxmax()
# Create a sub dataframe that only contains the information where life expectancy is most recent valid
recent_df = gapminder_data.loc[idx_max].set_index("name")
# Show the corresponding country, year, and life expectancy
recent_df[['time', 'lifeExpectancy']]

time  lifeExpectancy
name                                         
Afghanistan            2019.0        64.83300
Albania                2019.0        78.57300
Algeria                2019.0        76.88000
Angola                 2019.0        61.14700
Antigua and Barbuda    2019.0        77.01600
...                       ...             ...
Viet Nam               2019.0        75.40000
Virgin Islands (U.S.)  2019.0        79.66829
Yemen                  2019.0        66.12500
Zambia                 2019.0        63.88600
Zimbabwe               2019.0        61.49000

[204 rows x 2 columns]

    7. (7pt) Finally, let’s compute the growth rate. You can compute the growth rate (pct per year) as... 
    where LE0 is the life expectancy at the beginning of the period, LE1 is it at the end of the period,
    and n is the length of the period in years.
    Show the countries, drowth rates, and the corresponding life expectancies for 10 fastest and 10
    slowest growers.
    Hint: for each country, compute the first valid year of life expectancy, last valid year of life expectancy, and find their life expectancies for the corresponding years. You can do it in a fairly
    similar way as in the previous question. Now use this formula for each country. At the end just
    order the result.

In [31]:
# Create two series that contains the life expectancy of the most recent and the first valid
recent_le = recent_df.lifeExpectancy
valid_le = valid_df.lifeExpectancy
# Create two series that contains the year of the most recent LE and the first valid LE
recent_year = recent_df.time
valid_year = valid_df.time
# Calculate the growth rate using the formula
growth_rate = 100 * (((recent_le / valid_le) ** (1 / (recent_year - valid_year))) - 1)
# Drop the San Marino and Cayman Islands rows from the dataframe
growth_rate = growth_rate.drop(index = ['San Marino', 'Cayman Islands'], columns = 'name')
# Create two dataframes, one for the first valid year, one for most recent LE year row
first_le_df= pd.DataFrame({'Year': valid_year, 
                          'Life_Expectancy': valid_le, 
                          'Fertility_Rate': valid_df.fertilityRate})
recent_le_df = pd.DataFrame({'Year': recent_year, 
                          'Life_Expectancy': recent_le, 
                          'Fertility_Rate': recent_df.fertilityRate, 
                          'Growth_Rate': growth_rate})
# Combine the two dataframs together
present_df = pd.concat([first_le_df, recent_le_df], axis = 1)
# Show the 10 fastest growing countries
present_df.sort_values('Growth_Rate', ascending = False).head(10)

Year  Life_Expectancy  Fertility_Rate    Year  Life_Expectancy  \
name                                                                            
Yemen        1960.0           29.919           7.938  2019.0           66.125   
Maldives     1960.0           37.343           7.021  2019.0           78.921   
Mali         1960.0           28.199           6.967  2019.0           59.306   
Bhutan       1960.0           34.526           6.635  2019.0           71.777   
Timor-Leste  1960.0           33.729           6.319  2019.0           69.495   
Afghanistan  1960.0           32.446           7.450  2019.0           64.833   
Nepal        1960.0           35.583           5.959  2019.0           70.778   
Gambia       1960.0           32.054           6.246  2019.0           62.050   
Liberia      1960.0           34.264           6.406  2019.0           64.104   
Tunisia      1960.0           42.021           6.942  2019.0           76.699   

             Fertility_Rate  Growth_Rate  
name                                      
Yemen                 3.700     1.353232  
Maldives              1.840     1.276385  
Mali                  5.785     1.268013  
Bhutan                1.954     1.248151  
Timor-Leste           3.943     1.232786  
Afghanistan           4.321     1.180194  
Nepal                 1.876     1.172379  
Gambia                5.154     1.125813  
Liberia               4.247     1.067371  
Tunisia               2.174     1.025081

In [32]:
# Print the 10 slowest growing countries
present_df.sort_values('Growth_Rate').head(10)

Year  Life_Expectancy  Fertility_Rate    Year  \
name                                                                  
Marshall Islands    1987.0         72.14146            5.90  2000.0   
Palau               1990.0         69.06927            2.84  2005.0   
Ukraine             1960.0         68.29954            2.24  2019.0   
Latvia              1960.0         69.78683            1.94  2019.0   
Bulgaria            1960.0         69.24756            2.31  2019.0   
Lithuania           1960.0         69.84732            2.56  2019.0   
Seychelles          1980.0         69.59756             NaN  2019.0   
Belarus             1960.0         67.70810            2.67  2019.0   
Russian Federation  1960.0         66.05529            2.52  2019.0   
Slovakia            1960.0         69.92366            3.04  2019.0   

                    Life_Expectancy  Fertility_Rate  Growth_Rate  
name                                                              
Marshall Islands           65.23902             NaN    -0.770640  
Palau                      69.12927           2.020     0.005789  
Ukraine                    71.82732           1.228     0.085396  
Latvia                     75.23902           1.610     0.127581  
Bulgaria                   74.91463           1.580     0.133413  
Lithuania                  76.13415           1.610     0.146183  
Seychelles                 73.94390           2.340     0.155447  
Belarus                    74.22683           1.382     0.155918  
Russian Federation         73.08390           1.504     0.171530  
Slovakia                   77.47073           1.560     0.173873

In this case, San Marino and Cayman Islands are manually dropped. This is because San Marino and Cayman Islands only have life expectancy data for one year which means that the most recent year and the first year are the same. Thus, (most recent year - the first valid year) would equal to 0 which should not be included in the smallest growth rate consideration.

    8. (4pt) Do you see a pattern (or multiple patterns) here? Remember: you are looking at growth of life expectancy over an extended period.


It seems like all the countries except for Marshall Islands have a positive growth rate. It also seems like a majority of the slow growth countries have suffered certain kind of economic distruptions. I notice that the majority of the fastest growth countries started with a very low life expectancy. This indicates that even though their life expectancy today may not be competable with other countries, they still have higher growth rate because of the low starting life expectancy.

# References

Lubanovic, B. (2014) Introducing Python: Modern Computing in Simple Packages, O’Reilly Media.